modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionStateTracker.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.manager;
+
+import org.mule.extension.introspection.Configuration;
+import org.mule.extension.introspection.Extension;
+import org.mule.extension.introspection.Operation;
+import org.mule.extension.runtime.ConfigurationInstanceProvider;
+import org.mule.extension.runtime.OperationContext;
+import org.mule.extension.runtime.OperationExecutor;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+
+/**
+ * Holds state regarding the use that the platform is doing of a
+ * certain {@link Extension}
+ *
+ * @since 3.7.0
+ */
+final class ExtensionStateTracker
+{
+
+    private final LoadingCache<Configuration, ConfigurationStateTracker> configurationsStates = CacheBuilder.newBuilder()
+            .build(new CacheLoader<Configuration, ConfigurationStateTracker>()
+            {
+                @Override
+                public ConfigurationStateTracker load(Configuration configuration) throws Exception
+                {
+                    return new ConfigurationStateTracker();
+                }
+            });
+
+    <C> void registerConfigurationInstanceProvider(Configuration configuration,
+                                                   String providerName,
+                                                   ConfigurationInstanceProvider<C> configurationInstanceProvider)
+    {
+        getStateTracker(configuration).registerInstanceProvider(providerName, configurationInstanceProvider);
+    }
+
+    Map<String, ConfigurationInstanceProvider> getConfigurationInstanceProviders(Configuration configuration)
+    {
+        return ImmutableMap.copyOf(getStateTracker(configuration).getConfigurationInstanceProviders());
+    }
+
+    /**
+     * Registers a {@code configurationInstance} which is a realization of a {@link Configuration}
+     * model defined by {@code configuration}
+     *
+     * @param configuration         a {@link Configuration}
+     * @param instanceName          the name of the instance
+     * @param configurationInstance an instance which is compliant with the {@code configuration} model
+     * @param <C>                   the type of the configuration instance
+     */
+    <C> void registerConfigurationInstance(Configuration configuration, String instanceName, C configurationInstance)
+    {
+        getStateTracker(configuration).registerInstance(instanceName, configurationInstance);
```

Comment:
I'm a bit confused about this. Is not configurationInstance the extension instance?